### PR TITLE
Quantcast adaptor

### PR DESCRIPTION
**Pull request by ShreeniwasIyer**
We are implementing a new adaptor for Quantcast with a code name qcx for the bidderCode.

## Type of change

- [ ] Bugfix
- [ ] Feature
- [x] New bidder adapter
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other

## Description of change

- test parameters for validating bids
```
{
  bidder: 'quantcast',
  params: {
    publisherId: 'LCOf845vzU',
    battr : [1,2]
  }
}
```
- contact email of the adapter’s maintainer
- [Quantcast Inventory Supply Team] inventoryteam@quantcast.com

### CODE REVIEW

**Comment by aneuway2** [comment]
Hey @ShreeniwasIyer  are `publisherId` and `battr` both required parameters for your adaptor? Might be good to submit a PR to have this documented :-)

https://github.com/prebid/prebid.github.io/tree/master/dev-docs/bidders

**Comment by ShreeniwasIyer** [comment]
Thanks @aneuway2 for pointing out. Done now through https://github.com/prebid/prebid.github.io/pull/204

**Comment by ShreeniwasIyer** [comment]
Hi @aneuway2 , @mkendall07 - does this need any more inputs before review? I see that PRs submitted after this one have already been reviewed - so was wondering if I missed something..

**Comment by mkendall07** [comment]
@ShreeniwasIyer 
We will review this for the next release. Thanks

File: src/adapters/quantcast.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+const ajax = require('../ajax.js');
+const QUANTCAST_CALLBACK_URL = 'http://global.qc.rtb.quantserve.com:8080/qchb';
+
+var QuantcastAdapter = function QuantcastAdapter() {
+
+  const BIDDER_CODE = 'quantcast';
+
+  const DEFAULT_BID_FLOOR = 0.0000000001;
+  // The following 2 constants are adopted from bidfactory.js codes
+  const BID_STATUS_CODE_AVAILABLE = 1;
```
Comment (jaiminpanchal27) [inline src/adapters/quantcast.js]: Use STATUS.GOOD and STATUS.NO_BID defined in constants.json

File: src/adapters/quantcast.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+      }
+      return;
+    };
+
+  //expose the callback to the global object:
+  $$PREBID_GLOBAL$$.handleQuantcastCB = function (responseText) {
+    if(utils.isEmpty(responseText)) {
+      return;
+    }
+    let response = JSON.parse(responseText);
```
Comment (jaiminpanchal27) [inline src/adapters/quantcast.js]: wrap this in try catch

File: src/adapters/quantcast.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+    let referrer = utils.getTopWindowUrl();
+    let loc = utils.getTopWindowLocation();
+    let domain = loc.hostname;
+    let publisherId = 0;
+
+    publisherId = '' + bids[0].params.publisherId;
+    utils._each(bids, function(bid) {
+      // make sure the "sizes" are an array of arrays
+      if (!(bid.sizes[0] instanceof Array)) {
```
Comment (jaiminpanchal27) [inline src/adapters/quantcast.js]: DFP allows sizes to be array and array of arrays. Support both.
parseSizesInput function https://github.com/prebid/Prebid.js/blob/master/src/utils.js#L117 will help

File: src/adapters/quantcast.js [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  const DEFAULT_BID_FLOOR = 0.0000000001;
+  // The following 2 constants are adopted from bidfactory.js codes
+  const BID_STATUS_CODE_AVAILABLE = 1;
+  const BID_STATUS_CODE_EMPTY = 2;
+  let bidRequests = {};
+
+  let returnEmptyBid = function() {
+      var bidsRequested = $$PREBID_GLOBAL$$._bidsRequested.find(bidSet => bidSet.bidderCode === BIDDER_CODE).bids;
+      if (bidsRequested.length > 0) {
+        let bid = bidfactory.createBid(BID_STATUS_CODE_EMPTY);
```
Comment (jaiminpanchal27) [inline src/adapters/quantcast.js]: Pass bidRequest as second param

File: src/adapters/quantcast.js [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+  const BIDDER_CODE = 'quantcast';
+
+  const DEFAULT_BID_FLOOR = 0.0000000001;
+  // The following 2 constants are adopted from bidfactory.js codes
+  const BID_STATUS_CODE_AVAILABLE = 1;
+  const BID_STATUS_CODE_EMPTY = 2;
+  let bidRequests = {};
+
+  let returnEmptyBid = function() {
+      var bidsRequested = $$PREBID_GLOBAL$$._bidsRequested.find(bidSet => bidSet.bidderCode === BIDDER_CODE).bids;
```
Comment (jaiminpanchal27) [inline src/adapters/quantcast.js]: Use getBidRequest https://github.com/prebid/Prebid.js/blob/master/src/utils.js#L518

File: test/spec/adapters/quantcast_spec.js [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+          publisherId: 'test-publisher',
+          battr : [1,2],
+        }
+      }
+    ]
+  };
+
+  beforeEach(() => {
+    bidsRequestedOriginal = pbjs._bidsRequested;
+    pbjs._bidsRequested = [];
```
Comment (jaiminpanchal27) [inline test/spec/adapters/quantcast_spec.js]: Use $$PREBID_GLOBAL$$ at all places in this file

File: src/adapters/quantcast.js (hunk 1 above)
Comment (ShreeniwasIyer) [inline src/adapters/quantcast.js]: Done. Thanks. @jaiminpanchal27

**Comment by ShreeniwasIyer** [comment]
@jaiminpanchal27 I fixed the code as per the comments above. We are also currently pushing an change to our domain handling to allow for testing from localhost on certain test cases, this one inclusive. I believe it should be done in the next 3 hours or so.

**Comment by ShreeniwasIyer** [comment]
@jaiminpanchal27 All changes are done both in JS and server side and it seems we are good to go. Please review. (The last build is working from the PR perspective, but failing on some other test, which seems to pass sometimes and fail at other times. Please re-invoke the tests and I am confident they will pass.)

**Comment by jaiminpanchal27** [comment]
@ShreeniwasIyer I now see response but its not a valid bid. Here is the response i get.
`{"bidderCode":"qcx","bids":[{"height":250,"placementCode":"div-gpt-ad-1438287399331-0","statusCode":0,"width":300}],"requestId":"280f6c64a34097"}`

**Comment by ShreeniwasIyer** [comment]
Hi @jaiminpanchal27 , thanks for your review comments. We have fixed a few issues we noticed along the way. Now our bidder is definitely responding with the bids and would do so, so long as you are testing this on localhost and you are doing this from an US IP.

Can you please review and let us know?

**Comment by jaiminpanchal27** [comment]
@ShreeniwasIyer I still see the same CORS error. I am testing from US ip.
_Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at http://global.qc.rtb.quantserve.com:8080/qchb. (Reason: expected ‘true’ in CORS header ‘Access-Control-Allow-Credentials’).  (unknown)_

I am testing using this example page, https://jsfiddle.net/L618t79a/ Update prebid path before testing at your end. 

Ajax request to your endpoint
`curl 'http://global.qc.rtb.quantserve.com:8080/qchb' -H 'Cookie: mc=58f61b01-efd3a-4c41c-d26e7; d=EAUB4AEB4heBkwo9-Tk8LivvgwqLLx7NTB2MQsMAAAEPhSrqMLkgIgkgAAkQAAC-IADcMJiiKPURC6QAAAMNmGAQ6Vgri6KSuKwgDB7eEZ9Cn5LxnJVK1YLugfHI0YHeknji4f_bEO2i8f_ygpjH8ZzcIYsg' -H 'Origin: http://localhost:9999' -H 'Accept-Encoding: gzip, deflate' -H 'Accept-Language: en-US,en;q=0.8,gu;q=0.6' -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.81 Safari/537.36' -H 'Content-Type: text/plain' -H 'Accept: */*' -H 'Cache-Control: max-age=0' -H 'Referer: http://localhost:9999/integrationExamples/gpt/adapter/qcast.html' -H 'Connection: keep-alive' --data-binary '{"publisherId":"LCOf845vzU","requestId":"2c83a976870a59","bidId":"2c83a976870a59","site":{"page":"http://localhost:9999/integrationExamples/gpt/adapter/qcast.html","referrer":"http://localhost:9999/integrationExamples/gpt/adapter/qcast.html","domain":"localhost"},"imp":[{"banner":{"battr":[1,2],"sizes":[{"width":728,"height":90},{"width":300,"height":250},{"width":300,"height":600}]},"placementCode":"div-gpt-ad-1438287399331-0","bidFloor":1e-10}]}' --compressed`

Let me know if i am doing something wrong here.
You can also reach out to me on jpanchal@appnexus.com

**Comment by ShreeniwasIyer** [comment]
Thanks @jaiminpanchal27 for the review. Would you be merging it into the master when ready for next release? Do I need to do anything else?

**Comment by jaiminpanchal27** [comment]
@ShreeniwasIyer It looks good to me. No you don't need to do anything else from your side.
As per our process, any of the contributor will check and give second LGTM and merge.

File: src/adapters/quantcast.js [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    if(utils.isEmpty(responseText)) {
+      return;
+    }
+    let response = null;
+    try {
+      response = JSON.parse(responseText);
+    } catch(e) {
+      // Malformed JSON
+      utils.logError("Malformed JSON received from server - can't do anything here");
+      return;
```
Comment (mkendall07) [inline src/adapters/quantcast.js]: Should really try to add a error bid here if possible

Comment (ShreeniwasIyer) [inline src/adapters/quantcast.js]: @mkendall07 I tried to, but at that point in code we don't have the placement code for which the empty text response came and hence returning an error bid isn't possible..

**Comment by jaiminpanchal27** [comment]
This is merged into master. Please submit a PR to the [docs repo](https://github.com/prebid/prebid.github.io) to add a file for your adapter to the [bidders directory](https://github.com/prebid/prebid.github.io/tree/master/dev-docs/bidders) so your adapter's params will appear on the [bidders page](http://prebid.org/dev-docs/bidders.html). Thank you for contributing

**Pull request merged**